runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -921,20 +925,494 @@ static void VMwrtbarEvaluator(TR::Node *node, TR::Register *srcReg, TR::Register
+         {
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addis, node, scratchReg, cg->getTOCBaseRegister(), HI_VALUE(beginIndex));
+         generateTrg1MemInstruction(cg,TR::InstOpCode::Op_load, node, snippetReg, new (cg->trHeapMemory()) TR::MemoryReference(scratchReg, LO_VALUE(beginIndex), TR::Compiler->om.sizeofReferenceAddress(), cg));
+         }
+      else
+         {
+         generateTrg1MemInstruction(cg,TR::InstOpCode::Op_load, node, snippetReg, new (cg->trHeapMemory()) TR::MemoryReference(cg->getTOCBaseRegister(), beginIndex, TR::Compiler->om.sizeofReferenceAddress(), cg));   
+         }
+      }
+      
+   // Loaded Snippet
+   if (!isInstanceField)
+      {
+      static_cast<TR::J9PPCWatchedStaticFieldSnippet *>(dataSnippet)->loadSnippet();
+      }
+   else
+      {
+      static_cast<TR::J9PPCWatchedInstanceFieldSnippet *>(dataSnippet)->loadSnippet();
+      }
+   }
+
+TR::Snippet * 
+J9::Power::TreeEvaluator::getFieldWatchInstanceSnippet(TR::CodeGenerator *cg, TR::Node *node, J9Method *m, UDATA loc, UDATA os)
+   {
+   return new (cg->trHeapMemory()) TR::J9PPCWatchedInstanceFieldSnippet(cg, node, m, loc, os);
+   }
+
+TR::Snippet * 
+J9::Power::TreeEvaluator::getFieldWatchStaticSnippet(TR::CodeGenerator *cg, TR::Node *node, J9Method *m, UDATA loc, void *fieldAddress, J9Class *fieldClass)
+   {
+   return new (cg->trHeapMemory()) TR::J9PPCWatchedStaticFieldSnippet(cg, node, m, loc, fieldAddress, fieldClass);
    }
 
 void
-J9::Power::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister)
+J9::Power::TreeEvaluator::generateTestAndReportFieldWatchInstructions(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister, TR::Register *valueReg, TR::Register *dataSnippetRegister)
    {
-   TR_ASSERT_FATAL(false, "This helper implements platform specific code for Fieldwatch, which is currently not supported on Power platforms.\n");
+   bool is64Bit = TR::Compiler->target.is64Bit();
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   bool isInstanceField = node->getOpCode().isIndirect();
+   bool isAOTCompile = cg->comp()->compileRelocatableCode();
+   TR::Compilation *comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+
+   TR::Register *scratchReg = cg->allocateRegister();
+
+   // Check if snippet has been loaded
+   if (isInstanceField && !static_cast<TR::J9PPCWatchedInstanceFieldSnippet *>(dataSnippet)->isSnippetLoaded() ||
+      (!isInstanceField && !static_cast<TR::J9PPCWatchedStaticFieldSnippet *>(dataSnippet)->isSnippetLoaded()))
+      loadFieldWatchSnippet(cg, node, dataSnippet, dataSnippetRegister, scratchReg, isInstanceField);
+
+   TR::Register *fieldClassReg = cg->allocateRegister();
+   TR::MemoryReference *classFlagsMemRef = NULL;
+   TR::Register *cndReg = cg->allocateRegister(TR_CCR);
```

Comment:
Can we move some of these definitions down closer to where they are being used for the first time? Especially  `classFlagsMemRef` and `cndReg`.

Reply:
I agree on classFlagsMemRef, but I don't see the benefit on allocating cndReg where it's used. It's a bit easier to take account of all registers if they're allocated together.

Reply:
It limits the scope of the variable so we don't initialize it on paths in which we do not use it and it groups related code together. There are a ton more reasons to practice minimizing variable scope [1].

[1] https://www.google.com/search?q=minimizing+scope+of+variables

Reply:
To add on to @fjeremic's comment above, it also makes it easier to refactor/remove dead code if there are changes made in the area in the future.

Reply:
@AlenBadel have you made this change?

Reply:
Yes

Reply:
These have been addressed. Here are a few of the referenced examples:
https://github.com/eclipse/openj9/pull/6234/files#diff-efcdf3eda93e6ae70d4c46dd0bf9b5acR1324
https://github.com/eclipse/openj9/pull/6234/files#diff-efcdf3eda93e6ae70d4c46dd0bf9b5acR1002

I'm going to resolve this one. Please reopen if there are other concerns.

Reply:
I think this `cndReg` can be moved down much closer. Maybe even the `fieldClassReg`. Note that this is inside `generateTestAndReportFieldWatchInstructions`. 

Similar changes were made inside `generateFillInDataBlockSequenceForUnresolvedField`. So this one might have been missed.